### PR TITLE
Improve the presentation of validation results for automated signing

### PR DESCRIPTION
**Pull request by kmaglione**
I tried to break this up into three commits, but the changes to drop the saved escaped validation results became too entwined with the changes to add the `automated_signing` field, and a few other changes. I can probably break it up with some effort, though, if you think it's necessary.

### CODE REVIEW

File: apps/amo/urlresolvers.py [hunk 1]
```diff
@@ -214,6 +215,26 @@ def linkify_only_full_urls(attrs, new=False):
...
 
 
+URL_RE = re.compile(r'\bhttps?://([a-z0-9-]+.)+\.({0})/[^\s<>&"\x27]*'
+                    .format('|'.join(bleach.TLDS)))
+
+
+def linkify_escape(text):
+    """Linkifies plain text, escaping any HTML metacharacters already
+    present."""
+    # Bleach does a monumentally bad job at this. If we pass it # escaped HTML
```
Comment (kmaglione) [inline apps/amo/urlresolvers.py]: Yeah, I see...

File: apps/amo/urlresolvers.py [hunk 2]
```diff
@@ -214,6 +215,26 @@ def linkify_only_full_urls(attrs, new=False):
...
+URL_RE = re.compile(r'\bhttps?://([a-z0-9-]+\.)+({0})/[^\s<>&"\x27]*'
+                    .format('|'.join(bleach.TLDS)))
+
+
+def linkify_escape(text):
+    """Linkifies plain text, escaping any HTML metacharacters already
+    present."""
+    # Bleach does a monumentally bad job at this. If we pass it escaped HTML
+    # which contains any URLs (&lt;div&gt;http://foo.bar/&lt;div&gt;),
+    # we get back HTML (<div><a href="http://foo.bar/</div>).
```
Comment (mstriemer) [inline apps/amo/urlresolvers.py]: The HTML is malformed here. It just converts `&lt;div&gt;` to a `<div>` doesn't it?

File: apps/amo/urlresolvers.py [hunk 3]
```diff
@@ -214,6 +215,26 @@ def linkify_only_full_urls(attrs, new=False):
...
+    #
+    # So just stick to search-and-replace. We can hardly do a worse job than
+    # Bleach does.
+    def linkify(match):
+        # Parameters are already escaped.
+        return u'<a href="{0}">{0}</a>'.format(match.group(0))
```
Comment (mstriemer) [inline apps/amo/urlresolvers.py]: Don't we get bleach to put `rel="external"` on these?

File: static/js/zamboni/devhub.js [hunk 4]
```diff
@@ -104,7 +104,10 @@ $(document).ready(function() {
           $isSideload.attr('checked', false);
           $submitAddonProgress.removeClass('unlisted');
         } else {  // It's an unlisted add-on.
-          $uploadAddon.attr('data-upload-url', $uploadAddon.attr('data-upload-url-unlisted'));
+            if ($isSideload.is(':checked'))  // It's a sideload add-on, not eligible for automated signing.
```
Comment (mstriemer) [inline static/js/zamboni/devhub.js]: Indentation is off and this should use `{}`.

File: static/js/zamboni/validator.js [hunk 5]
```diff
@@ -144,6 +150,23 @@ function initValidator($doc) {
...
+        this.automatedSigning = suite.is('.automated-signing');
+        this.fileURL = url = suite.attr('data-file-url');
+
+        if (this.automatedSigning) {
+            this.hideNonSigning = $('#signing-hide-unnecessary').prop('checked');
+
+            // Hiding ignored messages only makes sense if we're only
+            // showing signing-related messages.
+            $('#signing-hide-ignored-container input').prop('disabled', !this.hideNonSigning);
+            if (!this.hideNonSigning)
```
Comment (mstriemer) [inline static/js/zamboni/validator.js]: `{}` here and below, too.

File: static/js/zamboni/validator.js [hunk 6]
```diff
@@ -194,71 +213,128 @@ function initValidator($doc) {
         return this.tiers[tierId];
     };
 
+    MsgVisitor.prototype.filterMessage = function(msg) {
+        if (this.hideIgnored && msg.ignored)
+            return false;
+
+        if (this.hideNonSigning && !msg.signing_severity)
```
Comment (mstriemer) [inline static/js/zamboni/validator.js]: What did `else` ever do to you? You could drop the `if`s and just return the boolean expression also.

```js
return !(this.hideIgnored && msg.ignored)
       && !(this.hideNonSigning && !msg.signing_severity);

File: apps/amo/urlresolvers.py (hunk 2 above)
Comment (kmaglione) [inline apps/amo/urlresolvers.py]: Sorry, typo in the original. It converts it to well-formed HTML, though: `<div><a href="http://foo.com/" rel="nofollow">http://foo.com/</a><div></div></div>`

I'll fix the comment.

File: apps/amo/urlresolvers.py (hunk 3 above)
Comment (kmaglione) [inline apps/amo/urlresolvers.py]: Yeah, normally, but these come from the validator, so they should be trusted URLs, and `rel="external"` is not helpful.

**Comment by magopian** [comment]
Separating commits helps a great deal for reviews. You could have made a separate commit for the removal of the `_escaped_validation` field, and another one for the revamping/refactoring of the views (and urls, and modifications to the templates).

Anyway, it was ok to review as is, thanks for the great work.
A few nits, some worries about modifications to the javascript and template code related to file submission (it's really intricate and often leads to errors, mistakes, oversights...), so it'll need a great deal of manual testing.

Rendering the validation results using (javascript or server-side) templates would be in a future PR, right? It makes sense, now that's you've already written the code to make it work as is.

**Comment by kmaglione** [comment]
Yeah, the problem is that I realized that the `_escaped_validation` field needed to be removed after I'd fixed the problems with the escaping (since the saved results weren't compatible with the rendering code anymore), and by the time I decided to split the PRs, there were just too many overlapping test changes. If I were to do it again, I'd have probably gone back and rebased a change to remove the field before fixing the escaping tests, but I missed that chance.

And, yeah, I was planning to submit another PR to rewrite using templates once I get feedback from UX about larger changes. These changes are kind of urgent, so I'd rather get a working version out as soon as possible.

**Comment by mstriemer** [comment]
Do you speed up how long it takes to escape validation results? It can take a really long time and was preventing add ons from being submitted before.

**Comment by kmaglione** [comment]
This version is almost certainly much faster, yes, but I don't think it actually matters. The most that function will be called during a submission should be twice, and even then, only if the dev looks at the full results page. And we haven't been caching escaped results for `FileValidation` objects, which hasn't caused any problems for the much heavier use that reviewers put it through.

**Comment by kmaglione** [comment]
I think this takes care of everything. I also made a couple of minor additional changes:

* Allow '.' and ',' in URLs for linkification, but ignore them if they come at the end of the URL.
* Link to the file compare rather than file browse link from the validation viewer when the results have been diffed against a previous version.

**Comment by magopian** [comment]
r+wc

**Pull request merged**